raiden/utils/migrations/v18_to_v19.py

```diff
@@ -0,0 +1,162 @@
+            # Get the state_change that triggered the event and if it has
+            # a block number get its hash. If not fall back to latest.
+            matched_state_changes = storage.get_statechanges_by_identifier(
+                from_identifier=event.state_change_identifier,
+                to_identifier=event.state_change_identifier,
+            )
+            result_length = len(matched_state_changes)
+            msg = 'multiple state changes should not exist for the same identifier'
+            assert result_length == 1, msg
+
+            statechange_data = json.loads(matched_state_changes[0])
+            if 'block_hash' in statechange_data:
+                block_hash = statechange_data['block_hash']
+            elif 'block_number' in statechange_data:
+                block_number = int(statechange_data['block_number'])
+                block_hash = cache.get(block_number)
+            data['triggered_by_block_hash'] = block_hash
+
+            updated_events.append(EventRecord(
+                event_identifier=event.event_identifier,
+                state_change_identifier=event.state_change_identifier,
+                data=json.dumps(data),
+            ))
+
+        storage.update_events(updated_events)
+
+
+def _transform_snapshot(
+        raw_snapshot: Dict[Any, Any],
+        storage: SQLiteStorage,
+        cache: BlockHashCache,
+) -> str:
+    """Upgrades a single snapshot by adding the blockhash to it and to any pending transactions"""
+    snapshot = json.loads(raw_snapshot)
+    block_number = int(snapshot['block_number'])
+    snapshot['block_hash'] = cache.get(block_number)
+
+    pending_transactions = snapshot['pending_transactions']
+    new_pending_transactions = []
+    for transaction_data in pending_transactions:
+        if 'raiden.transfer.events.ContractSend' not in transaction_data['_type']:
+            new_pending_transactions.append(transaction_data)
+            continue
+
+        # For each pending transaction find the corresponding DB event record.
+        event_record = storage.get_latest_event_by_data_field(
+            filters={'_type': transaction_data['_type']},
+        )
+        if not event_record.data:
+            raise InvalidDBData(
+                'Error during v18 -> v19 upgrade. Could not find a database event '
+                'table entry for a pending transaction.',
+            )
+
+        event_record_data = json.loads(event_record.data)
+        transaction_data['triggered_by_block_hash'] = event_record_data['triggered_by_block_hash']
+        new_pending_transactions.append(transaction_data)
+
+    snapshot['pending_transactions'] = new_pending_transactions
+    return json.dumps(snapshot)
+
+
+def _transform_snapshots_for_blockhash(storage: SQLiteStorage, cache: BlockHashCache) -> None:
+    """Upgrades the snapshots by adding the blockhash to it and to any pending transactions"""
+    updated_snapshots_data = []
+    for snapshot in storage.get_snapshots():
```

Comment:
This too can be a comprehension, and it can be done using a pool to parallelize execution.

Reply:
I tried to use gevent.Pool here too to parallelize. Commit is here: https://github.com/raiden-network/raiden/pull/3550/commits/45b04fd4ba6a746c855dd9340e53224663952bd0

But as in the other attempt to use gevent.Pool I would appreciate a second look here. I am not confident in my usage of it due to lack of using it before.